

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Analysis pipeline &mdash; RNA-HiC-tools 0.3.2 documentation</title>
  

  
  

  
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/labibi.css" type="text/css" />
  

  
    <link rel="top" title="RNA-HiC-tools 0.3.2 documentation" href="index.html"/>
        <link rel="next" title="Visualization of local RNA-RNA interactions" href="Visualization.html"/>
        <link rel="prev" title="RNA-Hi-C-tools 0.3 documentation" href="RNA-Hi-C-tools.html"/>


<script src="_static/labibi.js"></script>



<!-- Google Analytics JS is disabled; set google_analytics_id -->



  
  <script src="https://cdnjs.cloudflare.com/ajax/libs/modernizr/2.6.2/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        <a href="index.html" class="fa fa-home"> RNA-HiC-tools</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
        
            <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="RNA-Hi-C-tools.html">RNA-Hi-C-tools 0.3 documentation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="RNA-Hi-C-tools.html#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNA-Hi-C-tools.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="RNA-Hi-C-tools.html#support">Support</a></li>
</ul>
</li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Analysis pipeline</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#overview">Overview</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pipeline">Pipeline</a></li>
<li class="toctree-l2"><a class="reference internal" href="#other-functions">Other functions</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Visualization.html">Visualization of local RNA-RNA interactions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#prerequirement">Prerequirement</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#run-the-program-to-generate-visualization">Run the program to generate visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#example-of-result-graph">Example of result graph</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Visualization.html#visualization-of-intra-rna-interactions-by-heatmap">Visualization of intra-RNA interactions by heatmap</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#id1">Prerequirement</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#run-the-program-to-generate-heatmap-for-interactions-within-rna-molecule">Run the program to generate heatmap for interactions within RNA molecule</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#id2">Example of result graph</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Visualization.html#visualization-of-global-rna-rna-interactome">Visualization of global RNA-RNA interactome</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#id3">Prerequirement</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#id4">Run the program to generate visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#id5">Example of result graph</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Visualization.html#visualization-of-interaction-types-enrichment">Visualization of interaction types enrichment</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#id6">Prerequirement</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#run-the-program-to-generate-visualization-for-enrichment-of-different-types-of-interactions">Run the program to generate visualization for enrichment of different types of interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="Visualization.html#id7">Example of result graph</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Other_api.html">Python APIs created for this project</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Other_api.html#module-Annotation">Annotation module</a></li>
<li class="toctree-l2"><a class="reference internal" href="Other_api.html#annotated-bed-data-class">&#8220;annotated_bed&#8221; data class</a></li>
<li class="toctree-l2"><a class="reference internal" href="Other_api.html#rnastructure-class">&#8220;RNAstructure&#8221; class</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Data_Resources.html">Resources of strong interactions from two mouse cell types</a><ul>
<li class="toctree-l2"><a class="reference internal" href="Data_Resources.html#description-of-different-samples">Description of different samples</a></li>
<li class="toctree-l2"><a class="reference internal" href="Data_Resources.html#resources-of-strong-interactions">Resources of Strong Interactions</a></li>
<li class="toctree-l2"><a class="reference internal" href="Data_Resources.html#target-of-mirna-in-mir-290-295-clusters-and-mmu-mir-703">Target of miRNA in mir-290-295 clusters and mmu-mir-703</a></li>
</ul>
</li>
</ul>

        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">RNA-HiC-tools</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>Analysis pipeline</li>
      <li class="wy-breadcrumbs-aside">
        
          <a href="_sources/Analysis_pipeline.txt" rel="nofollow"> View page source</a>
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            
  
  <div class="section" id="analysis-pipeline">
<h1>Analysis pipeline<a class="headerlink" href="#analysis-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The next generation DNA sequencing based technology utilize RNA proximity ligation to transfrom RNA-RNA interactions into chimeric DNAs. Through sequencing and mapping these chimeric DNAs, it is able to achieve high-throughput mapping of nearly entire interaction networks. RNA linkers were introduced to mark the junction of the ligation and help to split the chimeric RNAs into two interacting RNAs.
This bioinformatic pipeline is trying to obtain the strong interactions from raw fastq sequencing data. The major steps are:</p>
<a class="reference internal image-reference" href="_images/pipeline.jpg"><img alt="_images/pipeline.jpg" class="align-center" src="_images/pipeline.jpg" style="width: 600px;" /></a>
<ul class="simple">
<li><a class="reference internal" href="#step1"><em>Step 1: Remove PCR duplicates.</em></a></li>
<li><a class="reference internal" href="#step2"><em>Step 2: Split library based on barcode.txt.</em></a></li>
<li><a class="reference internal" href="#step3"><em>Step 3: Recover fragments for each library.</em></a></li>
<li><a class="reference internal" href="#step4"><em>Step 4: Split partners and classify different types of fragments.</em></a></li>
<li><a class="reference internal" href="#step5"><em>Step 5: Align both parts of &#8220;Paired&#8221; fragment to the genome.</em></a></li>
<li><a class="reference internal" href="#step6"><em>Step 6: Determine strong interactions.</em></a></li>
<li><a class="reference internal" href="#step7"><em>Step 7: Visualization of interactions and coverages.</em></a></li>
</ul>
<p>Other functions:</p>
<ol class="arabic simple">
<li><a class="reference internal" href="#rna-types"><em>Determine the RNA types of different parts within fragments.</em></a></li>
<li><a class="reference internal" href="#find-linker"><em>Find linker sequences within the library.</em></a></li>
<li><a class="reference internal" href="#intersection"><em>Find intersections between two different interaction sets based on genomic locations</em></a></li>
<li><a class="reference internal" href="#intersectiongene"><em>Find intersections between two different interaction sets based on annotation</em></a></li>
<li><a class="reference internal" href="#structure"><em>RNA structure prediction by adding digestion site information</em></a></li>
<li><a class="reference internal" href="#snorna-mrna"><em>splicing intermediates detection within snoRNA-mRNA interactions</em></a></li>
</ol>
</div>
<div class="section" id="pipeline">
<h2>Pipeline<a class="headerlink" href="#pipeline" title="Permalink to this headline">¶</a></h2>
<div class="section" id="step-1-remove-pcr-duplicates">
<span id="step1"></span><h3>Step 1: Remove PCR duplicates.<a class="headerlink" href="#step-1-remove-pcr-duplicates" title="Permalink to this headline">¶</a></h3>
<p id="index-0">Starting from the raw pair-end sequencing data, PCR duplicates should be removed as the first step if both the 10nt random indexes and the remaining sequences are exactly the same for two pairs. It is achieved by <tt class="docutils literal"><span class="pre">remove_dup_PE.py</span></tt></p>
<div class="highlight-python"><pre>usage: remove_dup_PE.py [-h] reads1 reads2

Remove duplicated reads which have same sequences for both forward and reverse
reads. Choose the one appears first.

positional arguments:
  reads1      forward input fastq/fasta file
  reads2      reverse input fastq/fasta file

optional arguments:
  -h, --help  show this help message and exit

Library dependency: Bio, itertools</pre>
</div>
<p>The program will generate two fastq/fasta files after removind PCR duplicates and report how many read pairs has been removed. The output are prefixed with &#8216;Rm_dupPE&#8217;</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">One pair is considered as a PCR duplicate only when the sequences of both two ends (including the 10nt random index) are the exactly same as any of other pairs.</p>
</div>
</div>
<div class="section" id="step-2-split-library-based-on-barcode-txt">
<span id="step2"></span><h3>Step 2: Split library based on barcode.txt.<a class="headerlink" href="#step-2-split-library-based-on-barcode-txt" title="Permalink to this headline">¶</a></h3>
<p id="index-1">After removing PCR duplicates, the libraries from different samples are separated based on 4nt barcodes in the middle of random indexes (&#8220;RRRBBBBRRR&#8221;; R: random, B: barcode). It is implemented by <tt class="docutils literal"><span class="pre">split_library_pairend.py</span></tt></p>
<div class="highlight-python"><pre>usage: split_library_pairend.py [-h] [-f | -q] [-v] [-b BARCODE]
                                [-r RANGE [RANGE ...]] [-t] [-m MAX_SCORE]
                                input1 input2

Example: split_library_pairend.py -q Rm_dupPE_example.F1.fastq
         Rm_dupPE_example.R1.fastq -b barcode.txt

positional arguments:
  input1                input fastq/fasta file 1 for pairend data (contain
                        barcodes)
  input2                input fastq/fasta file 2 for pairend data

optional arguments:
  -h, --help            show this help message and exit
  -f, --fasta           add this option for fasta input file
  -q, --fastq           add this option for fastq input file
  -v, --version         show program's version number and exit
  -b BARCODE, --barcode BARCODE
                        barcode file
  -r RANGE [RANGE ...], --range RANGE [RANGE ...]
                        set range for barcode location within reads,default is
                        full read
  -t, --trim            trim sequence of 10nt index
  -m MAX_SCORE, --max_score MAX_SCORE
                        max(mismatch+indel) allowed for barcode match,
                        otherwise move reads into 'unassigned' file
                        default: 2.

Library dependency: Bio</pre>
</div>
<p>Here is a example for barcode.txt</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ACCT</span>
<span class="n">CCGG</span>
<span class="n">GGCG</span>
</pre></div>
</div>
<p>The output of this script are several pairs of fastq/fasta files prefixed with the 4nt barcode sequences, together with another pair of fastq/fasta files prefixed with &#8216;unassigned&#8217;.</p>
<p>For example, if the input fastq/fasta files are <tt class="docutils literal"><span class="pre">Rm_dupPE_example.F1.fastq</span></tt> and <tt class="docutils literal"><span class="pre">Rm_dupPE_example.R1.fastq</span></tt>, and the barcode file is the same as above, then the output files are:</p>
<ul class="simple">
<li>ACCT_Rm_dupPE_example.F1.fastq</li>
<li>ACCT_Rm_dupPE_example.R1.fastq</li>
<li>CCGG_Rm_dupPE_example.F1.fastq</li>
<li>CCGG_Rm_dupPE_example.R1.fastq</li>
<li>GGCG_Rm_dupPE_example.F1.fastq</li>
<li>GGCG_Rm_dupPE_example.R1.fastq</li>
<li>unassigned_Rm_dupPE_example.F1.fastq</li>
<li>unassigned_Rm_dupPE_example.R1.fastq</li>
</ul>
</div>
<div class="section" id="step-3-recover-fragments-for-each-library">
<span id="step3"></span><h3>Step 3: Recover fragments for each library.<a class="headerlink" href="#step-3-recover-fragments-for-each-library" title="Permalink to this headline">¶</a></h3>
<p id="index-2"><strong>After splitting the libraries, the later steps from here (Step 3-7) need to be executed parallelly for each sample.</strong></p>
<p>In this step, we are trying to recover the fragments based on local alignment. The fragments are classifed as several different types as shown in the figure below. The flow chart is also clarified at the top.</p>
<a class="reference internal image-reference" href="_images/workflow_for_recoverFragment.jpg"><img alt="_images/workflow_for_recoverFragment.jpg" class="align-center" src="_images/workflow_for_recoverFragment.jpg" style="width: 600px;" /></a>
<p>We will use a complied program <tt class="docutils literal"><span class="pre">recoverFragment</span></tt> to do that</p>
<div class="highlight-python"><pre>recoverFragment - recover fragment into 4 different categories from pair-end seq data
=====================================================================================

SYNOPSIS

DESCRIPTION
    -h, --help
          Displays this help message.
    --version
          Display version information
    -I, --inputs STR
          input of forward and reverse fastq file, path of two files separated by SPACE
    -p, --primer STR
          fasta file contianing two primer sequences
    -v, --verbose
          print alignment information for each alignment

EXAMPLES
    recoverFragment -I read_1.fastq read_2.fastq -p primer.fasta
          store fragment using fasta/fastq into 4 output files
          'short_*', 'long_*','evenlong_*','wierd_*'

VERSION
    recoverFragment version: 0.1
    Last update August 2013</pre>
</div>
</div>
<div class="section" id="step-4-split-partners-and-classify-different-types-of-fragments">
<span id="step4"></span><h3>Step 4: Split partners and classify different types of fragments.<a class="headerlink" href="#step-4-split-partners-and-classify-different-types-of-fragments" title="Permalink to this headline">¶</a></h3>
<p id="index-3">When we recovered the fragments, the next we are goting to do is to find RNA1 and RNA2 that are seprarated by the linkers, and from here, we will be able to classify the fragments into different types: &#8220;IndexOnly&#8221;, &#8220;NoLinker&#8221;, &#8220;LinkerOnly&#8221;, &#8220;BackOnly&#8221;, &#8220;FrontOnly&#8221;, &#8220;Paired&#8221;. (see the figure below).</p>
<a class="reference internal image-reference" href="_images/summary.jpg"><img alt="_images/summary.jpg" class="align-center" src="_images/summary.jpg" style="width: 600px;" /></a>
<p>This will be done by <tt class="docutils literal"><span class="pre">split_partner.py</span></tt></p>
<div class="highlight-python"><pre>usage: split_partner.py [-h] [-e EVALUE] [--linker_db LINKER_DB]
                        [--blast_path BLAST_PATH] [-o OUTPUT] [-t TRIM]
                        [-b BATCH] [-l LENGTH]
                        input type3_1 type3_2

DESCRIPTION: Run BLAST, find linker sequences and split two parts connected by
linkers

positional arguments:
  input                 the input fasta file containing fragment sequences of
                        type1 and type2
  type3_1               read_1 for evenlong (type3) fastq file
  type3_2               read_2 for evenlong (type3) fastq file

optional arguments:
  -h, --help            show this help message and exit
  -e EVALUE, --evalue EVALUE
                        cutoff evalues, only choose alignment with evalue less
                        than this cutoffs (default: 1e-5).
  --linker_db LINKER_DB
                        BLAST database of linker sequences
  --blast_path BLAST_PATH
                        path for the local blast program
  -o OUTPUT, --output OUTPUT
                        output file containing sequences of two sepatated
                        parts
  -t TRIM, --trim TRIM  trim off the first this number of nt as index,
                        default:10
  -b BATCH, --batch BATCH
                        batch this number of fragments for BLAST at a time.
                        default: 200000
  -r, --release         set to allow released criterion for Paired fragment in
                        Type 3, include those ones with no linker in two reads
  -l LENGTH, --length LENGTH
                        shortest length to be considered for each part of the
                        pair, default: 15

Library dependency: Bio, itertools</pre>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">New option added in version 0.3.1, which could allow two different strategies for selection of &#8220;Paired&#8221; fragments from the Type3 fragments. The <tt class="docutils literal"><span class="pre">--release</span></tt> option will allow a read pair to be called as &#8220;Paired&#8221; fragment even when the linker are not detected in both reads.</p>
</div>
<p>The linker fasta file contain sequences of all linkers</p>
<div class="highlight-python"><pre>&gt;L1
CTAGTAGCCCATGCAATGCGAGGA
&gt;L2
AGGAGCGTAACGTACCCGATGATC</pre>
</div>
<p>The output fasta files will be the input file name with different prefix (&#8220;NoLinker&#8221;, &#8220;LinkerOnly&#8221;, &#8220;BackOnly&#8221;, &#8220;FrontOnly&#8221;, &#8220;Paired&#8221;) for different types. The other output file specified by <tt class="docutils literal"><span class="pre">-o</span></tt> contains information of aligned linker sequences for each Type1/2 fragment.</p>
<p>For example, if the commend is</p>
<div class="highlight-python"><pre>split_partner.py fragment_ACCT.fasta evenlong_ACCTRm_dupPE_stitch_seq_1.fastq
    evenlong_ACCTRm_dupPE_stitch_seq_2.fastq
    -o fragment_ACCT_detail.txt --linker_db linker.fa</pre>
</div>
<dl class="docutils">
<dt>Then, the output files will be:</dt>
<dd><ul class="first last simple">
<li>backOnly_fragment_ACCT.fasta</li>
<li>NoLinker_fragment_ACCT.fasta</li>
<li>frontOnly_fragment_ACCT.fasta</li>
<li>Paired1_fragment_ACCT.fasta</li>
<li>Paired2_fragment_ACCT.fasta</li>
<li>fragment_ACCT_detail.txt</li>
</ul>
</dd>
</dl>
<p>The format of the last output file <tt class="docutils literal"><span class="pre">fragment_ACCT_detail.txt</span></tt> will be &#8220;Name | linker_num | linker_loc | Type | linker_order&#8221;. Here are two examples:</p>
<div class="highlight-python"><pre>HWI-ST1001:238:H0NYEADXX:1:1101:10221:1918      L1:2;L2:1  19,41;42,67;68,97       None    L2;L1;L1
HWI-ST1001:238:H0NYEADXX:1:1101:4620:2609       L1:2 28,46;47,79     Paired  L1;L1</pre>
</div>
<p>In the <strong>first</strong> fragment, there are three regions can be aligned to linkers, 2 for L1 and 1 for L2, the order is L2, L1, L1. And they are aligned in region [19,41], [42,67], [68,97] of the fragment. &#8220;None&#8221; means this fragment is either &#8216;LinkerOnly&#8217; or &#8216;IndexOnly&#8217; (in this case it is &#8216;LinkerOnly&#8217;). This fragment won&#8217;t be written to any of the output fasta files.</p>
<p>In the <strong>second</strong> fragment, two regions can be aligned to linkers, and they are both aligned to L1. The two regions are in [28,46], [47,79] of the fragment. the fragment is &#8220;Paired&#8221; because on both two sides flanking the linker aligned regions, the length is larger than 15nt. The left part will be writen in <tt class="docutils literal"><span class="pre">Paired1_fragment_ACCT.fasta</span></tt> and the right part in <tt class="docutils literal"><span class="pre">Paired2_fragment_ACCT.fasta</span></tt></p>
</div>
<div class="section" id="step-5-align-both-parts-of-paired-fragment-to-the-genome">
<span id="step5"></span><h3>Step 5: Align both parts of &#8220;Paired&#8221; fragment to the genome.<a class="headerlink" href="#step-5-align-both-parts-of-paired-fragment-to-the-genome" title="Permalink to this headline">¶</a></h3>
<p id="index-4">In this step, we will use the Paired1* and Paired2* fasta files output from the previous step. The sequences of part1 and part2 are aligned to the mouse genome mm9 with Bowtie and the pairs with both part1 and part2 mappable are selected as output. We also annotate the RNA types of each part in this step.
All of these are implemented using script <tt class="docutils literal"><span class="pre">Stitch-seq_Aligner.py</span></tt>.</p>
<div class="highlight-python"><pre>usage: Stitch-seq_Aligner.py [-h] [-s samtool_path] [-a ANNOTATION]
                             [-A DB_DETAIL]
                             miRNA_reads mRNA_reads bowtie_path miRNA_ref
                             mRNA_ref

Align miRNA-mRNA pairs for Stitch-seq. print the alignable miRNA-mRNA pairs
with coordinates

positional arguments:
  part1_reads           paired RNA1 fasta file
  part2_reads           paired RNA2 fasta file
  bowtie_path           path for the bowtie program
  part1_ref             reference genomic seq for RNA1
  part2_ref             reference genomic seq for RNA2

optional arguments:
  -h, --help            show this help message and exit
  -b, --bowtie2         set to use bowtie2 (--sensitive-local) for alignment,
                        need to change reference index and bowtie_path
  -u, --unique          set to only allow unique alignment
  -s samtool_path, --samtool_path samtool_path
                        path for the samtool program
  -a ANNOTATION, --annotation ANNOTATION
                        If specified, include the RNA type annotation for each
                        aligned pair, need to give bed annotation RNA file
  -A DB_DETAIL, --annotationGenebed DB_DETAIL
                        annotation bed12 file for lincRNA and mRNA with intron
                        and exon

Library dependency: Bio, pysam, itertools</pre>
</div>
<p>An annotation file for different types of RNAs in mm9 genome (bed format, &#8216;all_RNAs-rRNA_repeat.txt.gz&#8217;) was included in Data folder. The annotation bed12 file for lincRNA and mRNA (&#8216;Ensembl_mm9.genebed.gz&#8217;) was also included in Data folder. One can use the option <tt class="docutils literal"><span class="pre">-a</span> <span class="pre">../Data/all_RNAs-rRNA_repeat.txt.gz</span> <span class="pre">-A</span> <span class="pre">../Data/Ensembl_mm9.genebed.gz</span></tt> for annotation.</p>
<p>Here is a example:</p>
<div class="highlight-python"><pre>Stitch-seq_Aligner.py Paired1_fragment_ACCT.fasta Paired2_fragment_ACCT.fasta
    ~/Software/bowtie-0.12.7/bowtie mm9 mm9 -s samtools
    -a ../Data/all_RNAs-rRNA_repeat.txt.gz -A ../Data/Ensembl_mm9.genebed.gz
    &gt; ACCT_fragment_paired_align.txt</pre>
</div>
<p>The format for the output file <tt class="docutils literal"><span class="pre">ACCT_fragment_paired_align.txt</span></tt> will be:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="33%" />
<col width="68%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Column <a class="footnote-reference" href="#f1" id="id1">[1]</a></th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>chromosome name of RNA1</td>
</tr>
<tr class="row-odd"><td>2,3</td>
<td>start/end position of RNA1</td>
</tr>
<tr class="row-even"><td>4</td>
<td>strand information of RNA1</td>
</tr>
<tr class="row-odd"><td>5</td>
<td>sequence of RNA1</td>
</tr>
<tr class="row-even"><td>6</td>
<td>RNA type for RNA1</td>
</tr>
<tr class="row-odd"><td>7</td>
<td>RNA name for RNA1</td>
</tr>
<tr class="row-even"><td>8</td>
<td>RNA subtype <a class="footnote-reference" href="#f2" id="id2">[2]</a> for RNA1</td>
</tr>
<tr class="row-odd"><td>9</td>
<td>name of the pair</td>
</tr>
</tbody>
</table>
</div></blockquote>
<table class="docutils footnote" frame="void" id="f1" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id1">[1]</a></td><td>column 10-17 are the same as column 1-8 except they are for RNA2 instead of RNA1.</td></tr>
</tbody>
</table>
<table class="docutils footnote" frame="void" id="f2" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label"><a class="fn-backref" href="#id2">[2]</a></td><td>subtype can be intron/exon/utr5/utr3 for lincRNA and mRNA (protein-coding), &#8216;.&#8217; for others</td></tr>
</tbody>
</table>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Bowtie2 (&#8220;&#8211;sensitive-local&#8221; mode) option is added in version 0.3.1 for the user to choose, the <tt class="docutils literal"><span class="pre">reference</span> <span class="pre">index</span></tt> and <tt class="docutils literal"><span class="pre">bowtie_path</span></tt> need to be changed accordingly if you use bowtie2 instead of bowtie. User can also choose unique aligned reads or not by setting <tt class="docutils literal"><span class="pre">--unique</span></tt> option.</p>
</div>
</div>
<div class="section" id="step-6-determine-strong-interactions">
<span id="step6"></span><h3>Step 6: Determine strong interactions.<a class="headerlink" href="#step-6-determine-strong-interactions" title="Permalink to this headline">¶</a></h3>
<p id="index-5">In this step, we will generate clusters with high coverage separately for all RNA1 (R1) an RNA2 (R2) segments. Then based on the pairing information, we count the interactions between clusters from RNA1 and RNA2. For each interaction between clusters in RNA1 and RNA2, a p-value can be generated based on hypergeometric distribution. Given the p-values of all interactions, we could adjust the p-values controlled by False Discovery Rate (FDR, Benjamini-Hochberg procedure). The strong interactions can be selected by applying a FDR cutoff from adjusted p-values. (See figure below)</p>
<a class="reference internal image-reference" href="_images/Find_strong_interaction.jpg"><img alt="_images/Find_strong_interaction.jpg" class="align-center" src="_images/Find_strong_interaction.jpg" style="width: 600px;" /></a>
<p>We will use the script <tt class="docutils literal"><span class="pre">Select_strongInteraction_pp.py</span></tt>, parallel computing are implemented for clustering parallelly on different chromosomes:</p>
<div class="highlight-python"><pre>usage: Select_strongInteraction_pp.py [-h] -i INPUT [-M MIN_CLUSTERS]
                                      [-m MIN_INTERACTION] [-p P_VALUE]
                                      [-o OUTPUT] [-P PARALLEL] [-F]

find strong interactions from paired genomic location data

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT, --input INPUT
                        input file which is the output file of Stitch-seq-
                        Aligner.py
  -M MIN_CLUSTERS, --min_clusterS MIN_CLUSTERS
                        minimum number of segments allowed in each cluster,
                        default:5
  -m MIN_INTERACTION, --min_interaction MIN_INTERACTION
                        minimum number of interactions to support a strong
                        interaction, default:3
  -p P_VALUE, --p_value P_VALUE
                        the p-value based on hypergeometric distribution to
                        call strong interactions, default: 0.05
  -o OUTPUT, --output OUTPUT
                        specify output file
  -P PARALLEL, --parallel PARALLEL
                        number of workers for parallel computing, default: 5
  -F, --FDR             Compute FDR if specified

need Scipy for hypergeometric distribution</pre>
</div>
<p>The input of the script is the output of Step 5 (<tt class="docutils literal"><span class="pre">ACCT_fragment_paired_align.txt</span></tt> in the example). &#8220;annotated_bed&#8221; class is utilized in this script.</p>
<p>Here is a example:</p>
<div class="highlight-python"><pre>Select_strongInteraction_pp.py -i ACCT_fragment_paired_align.txt -o ACCT_interaction_clusters.txt</pre>
</div>
<p>The column description for output file <tt class="docutils literal"><span class="pre">ACCT_interaction_clusters.txt</span></tt> is:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="84%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Column</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>1</td>
<td>chromosome name of cluster in RNA1</td>
</tr>
<tr class="row-odd"><td>2,3</td>
<td>start/end position of cluster in RNA1</td>
</tr>
<tr class="row-even"><td>4</td>
<td>RNA type for cluster in RNA1</td>
</tr>
<tr class="row-odd"><td>5</td>
<td>RNA name for cluster in RNA1</td>
</tr>
<tr class="row-even"><td>6</td>
<td>RNA subtype for cluster in RNA1</td>
</tr>
<tr class="row-odd"><td>7</td>
<td># of counts for cluster in RNA1</td>
</tr>
<tr class="row-even"><td>8-14</td>
<td>Same as 1-7, but for cluster in RNA2</td>
</tr>
<tr class="row-odd"><td>15</td>
<td># of interactions between these two clusters</td>
</tr>
<tr class="row-even"><td>16</td>
<td>log(p-value) of the hypergeometric testing</td>
</tr>
</tbody>
</table>
</div></blockquote>
<p>We also have a set of scripts within the package to call strong interactions based on either clusters or RNA annotations, see the following table for detail:</p>
<blockquote>
<div><table border="1" class="docutils">
<colgroup>
<col width="31%" />
<col width="21%" />
<col width="48%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Call interaction based on</th>
<th class="head">Consider Left/Right</th>
<th class="head">Script</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Clusters (interaction sites)</td>
<td>Yes</td>
<td>Select_strongInteraction_pp.py</td>
</tr>
<tr class="row-odd"><td>Clusters (interaction sites)</td>
<td>No</td>
<td>Select_strongInteraction_pp_noLeftRight.py</td>
</tr>
<tr class="row-even"><td>RNA annotations</td>
<td>Yes</td>
<td>Select_strongInteraction_RNA.py</td>
</tr>
<tr class="row-odd"><td>RNA annotations</td>
<td>No</td>
<td>Select_strongInteraction_RNA_noLeftRight.py</td>
</tr>
</tbody>
</table>
</div></blockquote>
</div>
<div class="section" id="step-7-visualization-of-interactions-and-coverages">
<span id="step7"></span><h3>Step 7: Visualization of interactions and coverages.<a class="headerlink" href="#step-7-visualization-of-interactions-and-coverages" title="Permalink to this headline">¶</a></h3>
<p>There are two ways of visulization provided ( LOCAL and GLOBAL ):</p>
<blockquote>
<div><ul class="simple">
<li><a class="reference internal" href="Visualization.html#visualizationlocal"><em>Visualization of local interactions</em></a>.</li>
<li><a class="reference internal" href="Visualization.html#visualizationglobal"><em>Visualization of global interactome</em></a>.</li>
</ul>
</div></blockquote>
</div>
</div>
<div class="section" id="other-functions">
<h2>Other functions<a class="headerlink" href="#other-functions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="determine-the-rna-types-of-different-parts-within-fragments">
<span id="rna-types"></span><h3>Determine the RNA types of different parts within fragments.<a class="headerlink" href="#determine-the-rna-types-of-different-parts-within-fragments" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="find-linker-sequences-within-the-library">
<span id="find-linker"></span><h3>Find linker sequences within the library.<a class="headerlink" href="#find-linker-sequences-within-the-library" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="find-intersections-between-two-different-interaction-sets-based-on-genomic-locations">
<span id="intersection"></span><h3>Find intersections between two different interaction sets based on genomic locations<a class="headerlink" href="#find-intersections-between-two-different-interaction-sets-based-on-genomic-locations" title="Permalink to this headline">¶</a></h3>
<p id="index-6">The script tool <tt class="docutils literal"><span class="pre">intersectInteraction.py</span></tt> could be used to identify overlap of interactions between two interaction set from independent experiments based on genomic locations (two replicates or two different samples)</p>
<div class="highlight-python"><pre>usage: intersectInteraction.py [-h] -a FILEA -b FILEB [-s START] [-n NBASE]
                               [-o OUTPUT] [-c]

find intersections (overlaps) between two interaction sets

optional arguments:
  -h, --help            show this help message and exit
  -a FILEA, --filea FILEA
                        file for interaction set a
  -b FILEB, --fileb FILEB
                        file for interaction set b
  -s START, --start START
                        start column number of the second part in each
                        interaction (0-based), default:7
  -n NBASE, --nbase NBASE
                        number of overlapped nucleotides for each part of
                      interactions to call intersections, default: 1
  -o OUTPUT, --output OUTPUT
                        specify output file
  -p, --pvalue          calculate p-values based on 100times permutations

require 'random'&amp;'numpy'&amp;'scipy' module if set '-p'</pre>
</div>
<p>if &#8220;-p&#8221; option is set, then the program will do permutation for 100 times by shuffling the two partners of interactions in set a. A p-value will be calculate based on permutation distribution.</p>
</div>
<div class="section" id="find-intersections-between-two-different-interaction-sets-based-on-annotation">
<span id="intersectiongene"></span><h3>Find intersections between two different interaction sets based on annotation<a class="headerlink" href="#find-intersections-between-two-different-interaction-sets-based-on-annotation" title="Permalink to this headline">¶</a></h3>
<p id="index-7">The script tool <tt class="docutils literal"><span class="pre">intersectInteraction_genePair.R</span></tt> could be used to identify overlap of interactions between two interaction set from independent experiments based on the RNA annotations (two replicates or two different samples)</p>
<div class="highlight-python"><pre>usage: intersectInteraction_genePair.R [-h] [-n NUM [NUM ...]] [-p] [-r]
                                       [-o OUTPUT]
                                       interactionA interactionB

Call intersections based on gene pairs

positional arguments:
  interactionA          the interaction file a,[required]
  interactionB          the interaction file b,[required]

optional arguments:
  -h, --help            show this help message and exit
  -n NUM [NUM ...], --num NUM [NUM ...]
                        Column numbers for the gene name in two part,[default:
                        [5, 12]]
  -p, --pvalue          set to do 100 permutations for p-value of overlap
  -r, --release         set to only require match of chromosome and RNA name,
                        but not subtype
  -o OUTPUT, --output OUTPUT
                        output intersection file name, pairs in A that overlap
                        with B, [default: intersect.txt]</pre>
</div>
<p>if &#8220;-p&#8221; option is set, then the program will do permutation for 100 times by shuffling the two partners of interactions in both set a and set b. A p-value will be calculate based on permutation distribution.</p>
</div>
<div class="section" id="rna-structure-prediction-by-adding-digestion-site-information">
<span id="structure"></span><h3>RNA structure prediction by adding digestion site information<a class="headerlink" href="#rna-structure-prediction-by-adding-digestion-site-information" title="Permalink to this headline">¶</a></h3>
<p id="index-8">The script will take selfligated chimeric fragments from given snoRNA (ID) and predict secondary structures with and without constraints of digested single strand sites. It is also able to compare the known structure in dot format if the known structure is available and specified by &#8220;-a&#8221;. The script needs RNAStructure software for structure prediction (&#8220;-R&#8221;) and  and VARNA command line tool for visualization (&#8220;-v&#8221;).</p>
<div class="highlight-python"><pre>usage: RNA_structure_prediction.py [-h] [-g GENOMEFA] [-R RNASTRUCTUREEXE]
                                 [-a ACCEPTDOT] [-o OUTPUT]
                                 [-s samtool_path] [-v VARNA]
                                 [-c COLORMAPSTYLE]
                                 ID linkedPair

plot RNA structure with distribution of digested end, refine structure with
loc of digested end

positional arguments:
  ID                    Ensembl gene ID of RNA
  linkedPair            file for information of linked pairs, which is output
                        of 'Stitch-seq_Aligner.py'

optional arguments:
  -h, --help            show this help message and exit
  -g GENOMEFA, --genomeFa GENOMEFA
                        genomic sequence,need to be fadix-ed
  -R RNASTRUCTUREEXE, --RNAstructureExe RNASTRUCTUREEXE
                        folder of RNAstrucutre suite excutable
  -a ACCEPTDOT, --acceptDot ACCEPTDOT
                        accepted structure in dot format, for comparing of
                        accuracy, no comparison if not set
  -o OUTPUT, --output OUTPUT
                        output distribution of digested sites with dot
                        structures, can be format of eps, pdf, png,...
  -s samtool_path, --samtool_path samtool_path
                        path for the samtool program
  -v VARNA, --varna VARNA
                        path for the VARNA visualization for RNA
  -c COLORMAPSTYLE, --colorMapStyle COLORMAPSTYLE
                        style of color map, choose from: "red", "blue",
                        "green", "heat", "energy", and "bw",default:"heat"</pre>
</div>
<p>Here is a example:</p>
<div class="highlight-python"><pre>python RNA_structure_prediction.py \
  ENSMUSG00000064380 \
  /data2/sysbio/UCSD-sequencing/2013-11-27-Bharat_Tri_Shu/Undetermined_indices/Sample_lane8/ACCT_GGCG_combine/ACCT_GGCG_fragment_paired_align_selfLigation.txt \
  -a Snora73_real_dot.txt \
  -o Snora73_distribution.pdf</pre>
</div>
<p>Here &#8220;Snora73_real_dot.txt&#8221; is dot format of known Snora73 structure.
A example file for &#8220;Snora73_real_dot.txt&#8221;:</p>
<div class="highlight-python"><pre>&gt;Snora73
CCAACGUGGACAACCCAGGAGGUCACUCUCCCUGGGCUCUGUCCUAGUGGCAUAGGGGAGCAUAGGCCUUGCCCAGUGACGUACAGUCCCUUUCCACGGCGUUGGAGAUGAAGCUGGGCUUUGUGCCCGCGCCUGCAUAUUCCUACGACUUCUCAGAGUCCUGUGGACAGUGACUGAGGAGGCAAACCAUGCAGGAAACAGUG
((((((((((((((((.((((....))))....))))..)))))..((((.....(((((((((((.........))..)))))..))))...)))).)))))))..........................(((((((..........((((((...((((...)))).......)))))).......)))))))........</pre>
</div>
<p>The first line is the name of the small RNA, the second line is the RNA sequence and the third line is the dot format of secondary structure.</p>
<dl class="docutils">
<dt>This program will generate these files:</dt>
<dd><ul class="first last simple">
<li>Three eps files with secondary structures (&#8220;Predict&#8221;, &#8220;Refine&#8221;, &#8220;Accepted (known)&#8221;).</li>
<li>An output pdf file contains the distribution of digested sites in whole RNA molecule.</li>
<li>Two JSON files (&#8220;Predict&#8221;, &#8220;Refine&#8221;) to be uploaded into <a class="reference external" href="http://circos.zhu.land/">RNA2D-browser (Developed by Xiaopeng Zhu)</a> (using &#8220;read local file&#8221;)</li>
</ul>
</dd>
</dl>
<p>An example of the graph from eps file:</p>
<div class="figure align-center" style="width: 600px">
<a class="reference internal image-reference" href="_images/structure_ENSMUSG00000064380-SNORA73_cm_refine.jpg"><img alt="_images/structure_ENSMUSG00000064380-SNORA73_cm_refine.jpg" src="_images/structure_ENSMUSG00000064380-SNORA73_cm_refine.jpg" style="width: 600px;" /></a>
<p class="caption">The structure is predicted by &#8220;Fold&#8221; function of RNAstructure software and ploted by VARNA software. The distributions of RNase I digested sites across the RNA molecule are marked using a color scale on top of accepted snoRNA structures. The redder the nucleotide, the more frequently it could be digested by RNase I as suggested by our data. Non-base-paired regions tend to be more likely digested.</p>
</div>
<p>An example of the graph from pdf file with distribuiton of digested sites:</p>
<div class="figure align-center" style="width: 600px">
<a class="reference internal image-reference" href="_images/Snora73_digestedSite.jpg"><img alt="_images/Snora73_digestedSite.jpg" src="_images/Snora73_digestedSite.jpg" style="width: 600px;" /></a>
<p class="caption">legend: (A) Distribution of digested locations within the sequence of Snora73 RNA. The secondary structures were first predicted based only on the RNA sequences and then refined by adding information of single strand frequencies. The green regions are single stranded regions from the sequence based predicted structure. And the yellow regions are single stranded regions from the refined structure. Compared with the accepted structure (from fRNAdb), the positive predictive value is higher for the refined structure compared to the sequence based structure. The sensitivity is the same. (B, C) Counts of digested sites in the single stranded and double stranded portions of sequence-based predicted structure (B) and refined structure (C).</p>
</div>
<p>An example of graph generated by <a class="reference external" href="http://circos.zhu.land/">RNA2D-browser (Developed by Xiaopeng Zhu)</a> with the JSON file (PDF version cannot show this see HTML version)</p>
<iframe style="width:600px;height:600px;align:center" src="http://sky.zhu.land/cgi-bin/url2circos?url=http://systemsbio.ucsd.edu/RNA-Hi-C/_sources/structure_ENSMUSG00000064380-SNORA73.json"></iframe>
<p> The whole cirle shows the total length of SNORA73, the blue intensity histogram shows the distribution of RNase I digested locations. The dark grey links inside show the locations of different stems within the RNA structure. <\p></div>
<div class="section" id="splicing-intermediates-detection-within-snorna-mrna-interactions">
<span id="snorna-mrna"></span><h3>splicing intermediates detection within snoRNA-mRNA interactions<a class="headerlink" href="#splicing-intermediates-detection-within-snorna-mrna-interactions" title="Permalink to this headline">¶</a></h3>
<p id="index-9">This script is used to detect the percentage of potential splicing intermediates from all snoRNA-mRNA interactions within a sample. One snoRNA-mRNA interaction is considered as a potential splicing intermediate interaction if: <strong>snoRNA gene is located in the intronic region of its mRNA interacting partner.</strong> The script need an annotation bed file (<em>all_RNAs-rRNA_repeat.txt</em>):</p>
<div class="highlight-python"><pre>usage: snoRNA_mRNA_statistics.py [-h] [-A ANNOTATION] [-s START] [-o OUTPUT]
                               interaction

Statistics for snoRNA-mRNA interactions, percentage of splicing intermediates
within all snoRNA-mRNA interactions

positional arguments:
  interaction           Interaction file from output of
                        'Select_strongInteraction_pp.py', or linked fragment
                        pair file from output of 'Stitch-seq_Aligner.py'

optional arguments:
  -h, --help            show this help message and exit
  -A ANNOTATION, --Annotation ANNOTATION
                        Annotation bed file for locations of mRNA genes
  -s START, --start START
                        start column number of the second region in
                        interaction file, default=7
  -o OUTPUT, --output OUTPUT
                        Set to output all snoRNA-mRNA interactions as splicing
                        intermediates, not output if not set

Require: xplib, subprocess</pre>
</div>
<p>Here is a example:</p>
<div class="highlight-python"><pre>python snoRNA_mRNA_statistics.py \
  /data2/projects/rnarna/2013-11-RNA-RNA/ACCT_GGCG_combine/ACCT_GGCG_interaction_clusters_rmrRNA.txt \
   -A ../Data/all_RNAs-rRNA_repeat.txt.gz</pre>
</div>
<p>The output will be:</p>
<div class="highlight-python"><pre>snoRNA-mRNA-interactions:       8043
SplicingIntermediates:  4
Percentage      0.0497%</pre>
</div>
<p>If <tt class="docutils literal"><span class="pre">-o</span></tt> is set with a file name, then the 4 splicing intermediates like interactions will be output to that file.</p>
<p>The input can also be an linked fragment pair file, see this example:</p>
<div class="highlight-python"><pre>python snoRNA_mRNA_statistics.py \
  /data2/projects/rnarna/GGCG_combine_MEF_1/GGCG_fragment_paired_align_rmSingleFragment.txt \
  -s 9 \
  -A ../Data/all_RNAs-rRNA_repeat.txt.gz</pre>
</div>
<p>The output print is:</p>
<div class="highlight-python"><pre>snoRNA-mRNA-linkedPair: 24138
SplicingIntermediates:  19
Percentage      0.0787%</pre>
</div>
</div>
</div>
</div>




  

<div id="disqus_thread"></div>
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'stitchseqtools'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'http://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>



          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Visualization.html" class="btn btn-neutral float-right" title="Visualization of local RNA-RNA interactions"/>Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="RNA-Hi-C-tools.html" class="btn btn-neutral" title="RNA-Hi-C-tools 0.3 documentation"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2014, Zhong lab.
    </p>
  </div>

  <a href="https://github.com/snide/sphinx_rtd_theme">Sphinx theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>
</footer>
        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'0.3.2',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
  
 



<div id="editor-trap">
      <h3>Edit this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/yu68/RNA-Hi-C/blob/master/Analysis_pipeline.rst">
          Analysis pipeline
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/yu68/RNA-Hi-C/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>



<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', ]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>



</body>
</html>