<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Analysis pipeline &mdash; Stitch-seq-tools 1.0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Stitch-seq-tools 1.0.1 documentation" href="index.html" />
    <link rel="prev" title="Stitch-seq-tools 1.0 documentation" href="Stitch-seq-tools.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="Stitch-seq-tools.html" title="Stitch-seq-tools 1.0 documentation"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">Stitch-seq-tools 1.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="analysis-pipeline">
<h1>Analysis pipeline<a class="headerlink" href="#analysis-pipeline" title="Permalink to this headline">¶</a></h1>
<div class="section" id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>The next generation DNA sequencing based technology utilize RNA proximity ligation to transfrom RNA-RNA interactions into chimeric DNAs. Through sequencing and mapping these chimeric DNAs, it is able to achieve high-throughput mapping of nearly entire interaction networks. RNA linkers were introduced to mark the junction of the ligation and help to split the chimeric RNAs into two interacting RNAs.
This bioinformatic pipeline is trying to obtain the strong interactions from raw fastq sequencing data. The major steps are:</p>
<ul class="simple">
<li><a class="reference internal" href="#step1"><em>Step 1: Remove PCR duplicates.</em></a></li>
<li><a class="reference internal" href="#step2"><em>Step 2: Split library based on barcode.txt.</em></a></li>
<li><a class="reference internal" href="#step3"><em>Step 3: Recover fragments for each library.</em></a></li>
<li><a class="reference internal" href="#step4"><em>Step 4: Split partners and classify different types of fragments.</em></a></li>
<li><a class="reference internal" href="#step5"><em>Step 5: Align both parts of &#8220;Paired&#8221; fragment to the genome.</em></a></li>
<li><a class="reference internal" href="#step6"><em>Step 6: Determine strong interactions.</em></a></li>
</ul>
<p>Other functions:</p>
<ol class="arabic simple">
<li><a class="reference internal" href="#rna-types"><em>Determine the RNA types of different parts within fragments.</em></a></li>
<li><a class="reference internal" href="#find-linker"><em>Find linker sequences within the library.</em></a></li>
</ol>
</div>
<div class="section" id="pipeline">
<h2>Pipeline<a class="headerlink" href="#pipeline" title="Permalink to this headline">¶</a></h2>
<div class="section" id="step-1-remove-pcr-duplicates">
<span id="step1"></span><h3>Step 1: Remove PCR duplicates.<a class="headerlink" href="#step-1-remove-pcr-duplicates" title="Permalink to this headline">¶</a></h3>
<p>Starting from the raw pair-end sequencing data, PCR duplicates should be removed as the first step if both the 10nt random indexes and the remaining sequences are exactly the same for two pairs. It is achieved by <tt class="docutils literal"><span class="pre">remove_dup_PE.py</span></tt></p>
<div class="highlight-python"><pre>usage: remove_dup_PE.py [-h] reads1 reads2

Remove duplicated reads which have same sequences for both forward and reverse
reads. Choose the one appears first.

positional arguments:
  reads1      forward input fastq/fasta file
  reads2      reverse input fastq/fasta file

optional arguments:
  -h, --help  show this help message and exit

Library dependency: Bio, itertools</pre>
</div>
<p>The program will generate two fastq/fasta files after removind PCR duplicates and report how many read pairs has been removed. The output are prefixed with &#8216;Rm_dupPE&#8217;</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">One pair is considered as a PCR duplicate only when the sequences of both two ends (including the 10nt random index) are the exactly same as any of other pairs.</p>
</div>
</div>
<div class="section" id="step-2-split-library-based-on-barcode-txt">
<span id="step2"></span><h3>Step 2: Split library based on barcode.txt.<a class="headerlink" href="#step-2-split-library-based-on-barcode-txt" title="Permalink to this headline">¶</a></h3>
<p>After removing PCR duplicates, the libraries from different samples are separated based on 4nt barcodes in the middle of random indexes (&#8220;RRRBBBBRRR&#8221;; R: random, B: barcode). It is implemented by <tt class="docutils literal"><span class="pre">split_library_pairend.py</span></tt></p>
<div class="highlight-python"><pre>usage: split_library_pairend.py [-h] [-f | -q] [-v] [-b BARCODE]
                                [-r RANGE [RANGE ...]] [-t] [-m MAX_SCORE]
                                input1 input2

Example: split_library_pairend.py -q Rm_dupPE_example.F1.fastq
         Rm_dupPE_example.R1.fastq -b barcode.txt

positional arguments:
  input1                input fastq/fasta file 1 for pairend data (contain
                        barcodes)
  input2                input fastq/fasta file 2 for pairend data

optional arguments:
  -h, --help            show this help message and exit
  -f, --fasta           add this option for fasta input file
  -q, --fastq           add this option for fastq input file
  -v, --version         show program's version number and exit
  -b BARCODE, --barcode BARCODE
                        barcode file
  -r RANGE [RANGE ...], --range RANGE [RANGE ...]
                        set range for barcode location within reads,default is
                        full read
  -t, --trim            trim sequence of 10nt index
  -m MAX_SCORE, --max_score MAX_SCORE
                        max(mismatch+indel) allowed for barcode match,
                        otherwise move reads into 'unassigned' file
                        default: 2.

Library dependency: Bio</pre>
</div>
<p>Here is a example for barcode.txt</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ACCT</span>
<span class="n">CCGG</span>
<span class="n">GGCG</span>
</pre></div>
</div>
<p>The output of this script are several pairs of fastq/fasta files prefixed with the 4nt barcode sequences, together with another pair of fastq/fasta files prefixed with &#8216;unassigned&#8217;.</p>
<p>For example, if the input fastq/fasta files are <tt class="docutils literal"><span class="pre">Rm_dupPE_example.F1.fastq</span></tt> and <tt class="docutils literal"><span class="pre">Rm_dupPE_example.R1.fastq</span></tt>, and the barcode file is the same as above, then the output files are</p>
<div class="highlight-python"><div class="highlight"><pre><span class="n">ACCT_Rm_dupPE_example</span><span class="o">.</span><span class="n">F1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">ACCT_Rm_dupPE_example</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">CCGG_Rm_dupPE_example</span><span class="o">.</span><span class="n">F1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">CCGG_Rm_dupPE_example</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">GGCG_Rm_dupPE_example</span><span class="o">.</span><span class="n">F1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">GGCG_Rm_dupPE_example</span><span class="o">.</span><span class="n">R1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">unassigned_Rm_dupPE_example</span><span class="o">.</span><span class="n">F1</span><span class="o">.</span><span class="n">fastq</span>
<span class="n">unassigned_Rm_dupPE_example</span><span class="o">.</span><span class="n">F1</span><span class="o">.</span><span class="n">fastq</span>
</pre></div>
</div>
</div>
<div class="section" id="step-3-recover-fragments-for-each-library">
<span id="step3"></span><h3>Step 3: Recover fragments for each library.<a class="headerlink" href="#step-3-recover-fragments-for-each-library" title="Permalink to this headline">¶</a></h3>
<p><strong>After splitting the libraries, the later steps from here (Step 3-6) are executed parallelly for each sample.</strong></p>
<p>In this step, we are trying to recover the fragments based on local alignment. The fragments are classifed as several different types as shown in the figure below. The flow chart is also clarified at the top.</p>
<a class="reference internal image-reference" href="_images/workflow_for_recoverFragment.jpg"><img alt="_images/workflow_for_recoverFragment.jpg" src="_images/workflow_for_recoverFragment.jpg" style="width: 600px;" /></a>
<p>We will use a complied program <tt class="docutils literal"><span class="pre">recoverFragment</span></tt> to do that</p>
<div class="highlight-python"><pre>recoverFragment - recover fragment into 4 different categories from pair-end seq data
=====================================================================================

SYNOPSIS

DESCRIPTION
    -h, --help
          Displays this help message.
    --version
          Display version information
    -I, --inputs STR
          input of forward and reverse fastq file, path of two files separated by SPACE
    -p, --primer STR
          fasta file contianing two primer sequences
    -v, --verbose
          print alignment information for each alignment

EXAMPLES
    recoverFragment -I read_1.fastq read_2.fastq -p primer.fasta
          store fragment using fasta/fastq into 4 output files
          'short_*', 'long_*','evenlong_*','wierd_*'

VERSION
    recoverFragment version: 0.1
    Last update August 2013</pre>
</div>
</div>
<div class="section" id="step-4-split-partners-and-classify-different-types-of-fragments">
<span id="step4"></span><h3>Step 4: Split partners and classify different types of fragments.<a class="headerlink" href="#step-4-split-partners-and-classify-different-types-of-fragments" title="Permalink to this headline">¶</a></h3>
<p>When we recovered the fragments, the next we are goting to do is to find parts that are seprarated by the linkers, and from here, we will be able to classify the fragments into different types: &#8220;IndexOnly&#8221;, &#8220;NoLinker&#8221;, &#8220;LinkerOnly&#8221;, &#8220;BackOnly&#8221;, &#8220;FrontOnly&#8221;, &#8220;Paired&#8221;. (see the figure below).</p>
<a class="reference internal image-reference" href="_images/summary.jpg"><img alt="_images/summary.jpg" src="_images/summary.jpg" style="width: 600px;" /></a>
<p>This will be done by <tt class="docutils literal"><span class="pre">split_partner.py</span></tt></p>
<div class="highlight-python"><pre>usage: split_partner.py [-h] [-e EVALUE] [--linker_db LINKER_DB]
                        [--blast_path BLAST_PATH] [-o OUTPUT] [-t TRIM]
                        [-b BATCH] [-l LENGTH]
                        input type3_1 type3_2

DESCRIPTION: Run BLAST, find linker sequences and split two parts connected by
linkers

positional arguments:
  input                 the input fasta file containing fragment sequences of
                        type1 and type2
  type3_1               read_1 for evenlong (type3) fastq file
  type3_2               read_2 for evenlong (type3) fastq file

optional arguments:
  -h, --help            show this help message and exit
  -e EVALUE, --evalue EVALUE
                        cutoff evalues, only choose alignment with evalue less
                        than this cutoffs (default: 1e-5).
  --linker_db LINKER_DB
                        BLAST database of linker sequences
  --blast_path BLAST_PATH
                        path for the local blast program
  -o OUTPUT, --output OUTPUT
                        output file containing sequences of two sepatated
                        parts
  -t TRIM, --trim TRIM  trim off the first this number of nt as index,
                        default:10
  -b BATCH, --batch BATCH
                        batch this number of fragments for BLAST at a time.
                        default: 100000
  -l LENGTH, --length LENGTH
                        shortest length to be considered for each part of the
                        pair, default: 15

Library dependency: Bio, itertools</pre>
</div>
<p>The linker fasta file contain sequences of all linkers</p>
<div class="highlight-python"><pre>&gt;L1
CTAGTAGCCCATGCAATGCGAGGA
&gt;L2
AGGAGCGTAACGTACCCGATGATC</pre>
</div>
<p>The output fasta files will be the input file name with different prefix (&#8220;NoLinker&#8221;, &#8220;LinkerOnly&#8221;, &#8220;BackOnly&#8221;, &#8220;FrontOnly&#8221;, &#8220;Paired&#8221;) for different types. The other output file specified by <tt class="docutils literal"><span class="pre">-o</span></tt> contains information of aligned linker sequences for each Type1/2 fragment.</p>
<p>For example, if the commend is</p>
<div class="highlight-python"><pre>split_partner.py fragment_ACCT.fasta evenlong_ACCTRm_dupPE_stitch_seq_1.fastq evenlong_ACCTRm_dupPE_stitch_seq_2.fastq -o fragment_ACCT_detail.txt --linker_db linker.fa</pre>
</div>
<dl class="docutils">
<dt>Then, the output files will be:</dt>
<dd><ul class="first last simple">
<li>backOnly_fragment_ACCT.fasta</li>
<li>NoLinker_fragment_ACCT.fasta</li>
<li>frontOnly_fragment_ACCT.fasta</li>
<li>Paired1_fragment_ACCT.fasta</li>
<li>Paired2_fragment_ACCT.fasta</li>
<li>fragment_ACCT_detail.txt</li>
</ul>
</dd>
</dl>
<p>The format of the last output file <tt class="docutils literal"><span class="pre">fragment_ACCT_detail.txt</span></tt> will be &#8220;Name|linker_num|linker_loc|Type|linker_order&#8221;. Here are two examples:</p>
<div class="highlight-python"><pre>HWI-ST1001:238:H0NYEADXX:1:1101:10221:1918      L1:2;L2:1  19,41;42,67;68,97       None    L2;L1;L1
HWI-ST1001:238:H0NYEADXX:1:1101:4620:2609       L1:2 28,46;47,79     Paired  L1;L1</pre>
</div>
<p>In the <strong>first</strong> fragment, there are three regions can be aligned to linkers, 2 for L1 and 1 for L2, the order is L2, L1, L1. And they are aligned in region [19,41], [42,67], [68,97] of the fragment. <tt class="docutils literal"><span class="pre">None</span></tt> means this fragment is either &#8216;LinkerOnly&#8217; or &#8216;IndexOnly&#8217; (in this case it is &#8216;LinkerOnly&#8217;). This fragment won&#8217;t be written to any of the output fasta files.</p>
<p>In the <strong>second</strong> fragment, two regions can be aligned to linkers, and they are both aligned to L1. The two regions are in [28,46], [47,79] of the fragment. the fragmen is &#8220;Paired&#8221; because on both two sides flanking the linker aligned regions, the length is larger than 15nt. The left part will be writen in <tt class="docutils literal"><span class="pre">Paired1_fragment_ACCT.fasta</span></tt> and the right part in <tt class="docutils literal"><span class="pre">Paired2_fragment_ACCT.fasta</span></tt></p>
</div>
<div class="section" id="step-5-align-both-parts-of-paired-fragment-to-the-genome">
<span id="step5"></span><h3>Step 5: Align both parts of &#8220;Paired&#8221; fragment to the genome.<a class="headerlink" href="#step-5-align-both-parts-of-paired-fragment-to-the-genome" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="step-6-determine-strong-interactions">
<span id="step6"></span><h3>Step 6: Determine strong interactions.<a class="headerlink" href="#step-6-determine-strong-interactions" title="Permalink to this headline">¶</a></h3>
</div>
</div>
<div class="section" id="other-functions">
<h2>Other functions<a class="headerlink" href="#other-functions" title="Permalink to this headline">¶</a></h2>
<div class="section" id="determine-the-rna-types-of-different-parts-within-fragments">
<span id="rna-types"></span><h3>Determine the RNA types of different parts within fragments.<a class="headerlink" href="#determine-the-rna-types-of-different-parts-within-fragments" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="find-linker-sequences-within-the-library">
<span id="find-linker"></span><h3>Find linker sequences within the library.<a class="headerlink" href="#find-linker-sequences-within-the-library" title="Permalink to this headline">¶</a></h3>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Analysis pipeline</a><ul>
<li><a class="reference internal" href="#overview">Overview</a></li>
<li><a class="reference internal" href="#pipeline">Pipeline</a><ul>
<li><a class="reference internal" href="#step-1-remove-pcr-duplicates">Step 1: Remove PCR duplicates.</a></li>
<li><a class="reference internal" href="#step-2-split-library-based-on-barcode-txt">Step 2: Split library based on barcode.txt.</a></li>
<li><a class="reference internal" href="#step-3-recover-fragments-for-each-library">Step 3: Recover fragments for each library.</a></li>
<li><a class="reference internal" href="#step-4-split-partners-and-classify-different-types-of-fragments">Step 4: Split partners and classify different types of fragments.</a></li>
<li><a class="reference internal" href="#step-5-align-both-parts-of-paired-fragment-to-the-genome">Step 5: Align both parts of &#8220;Paired&#8221; fragment to the genome.</a></li>
<li><a class="reference internal" href="#step-6-determine-strong-interactions">Step 6: Determine strong interactions.</a></li>
</ul>
</li>
<li><a class="reference internal" href="#other-functions">Other functions</a><ul>
<li><a class="reference internal" href="#determine-the-rna-types-of-different-parts-within-fragments">Determine the RNA types of different parts within fragments.</a></li>
<li><a class="reference internal" href="#find-linker-sequences-within-the-library">Find linker sequences within the library.</a></li>
</ul>
</li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Stitch-seq-tools.html"
                        title="previous chapter">Stitch-seq-tools 1.0 documentation</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/Analysis_pipeline.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="Stitch-seq-tools.html" title="Stitch-seq-tools 1.0 documentation"
             >previous</a> |</li>
        <li><a href="index.html">Stitch-seq-tools 1.0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2013, Pengfei Yu.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b2.
    </div>
  </body>
</html>