<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Visualization of RNA-RNA interactome &mdash; RNA-Stitch-seq-tools 0.2.1 documentation</title>
    
    <link rel="stylesheet" href="_static/labibi.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.2.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="RNA-Stitch-seq-tools 0.2.1 documentation" href="index.html" />
    <link rel="next" title="Python APIs created for this project" href="Other_api.html" />
    <link rel="prev" title="Analysis pipeline" href="Analysis_pipeline.html" />


<script src="_static/labibi.js"></script>



<!-- Google Analytics JS is disabled; set google_analytics_id -->


  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Other_api.html" title="Python APIs created for this project"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Analysis_pipeline.html" title="Analysis pipeline"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">RNA-Stitch-seq-tools 0.2.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  
  <div class="section" id="visualization-of-rna-rna-interactome">
<span id="visualization"></span><h1>Visualization of RNA-RNA interactome<a class="headerlink" href="#visualization-of-rna-rna-interactome" title="Permalink to this headline">¶</a></h1>
<div class="section" id="prerequirement">
<h2>Prerequirement<a class="headerlink" href="#prerequirement" title="Permalink to this headline">¶</a></h2>
<p>This program is powered by <a class="reference external" href="http://cran.r-project.org/web/packages/RCircos/index.html">RCircos</a>.</p>
<p>Required R packages (our program will check for the presence of these packages and install/load them automatically if not present):</p>
<blockquote>
<div><ul class="simple">
<li>argparse, RCircos, biovizBase, rtracklayer</li>
</ul>
</div></blockquote>
<p>The program also require a python script &#8220;bam2tab.py&#8221; (already in /bin/ folder) to call coverage from <a class="reference external" href="https://github.com/nimezhu/bam2x/blob/master/scripts/bed2tab.py">BAM2X</a></p>
</div>
<div class="section" id="run-the-program-to-generate-visualization">
<h2>Run the program to generate visualization<a class="headerlink" href="#run-the-program-to-generate-visualization" title="Permalink to this headline">¶</a></h2>
<p id="index-0">We will use the script &#8220;Plot_Circos.R&#8221; for this purpose.</p>
<div class="highlight-python"><pre>usage: Plot_Circos.R [-h] [-g GENOME] [-b BIN] [-o OUTPUT]
                   interaction part1 part2

positional arguments:
  interaction           the interaction file,[required]
  part1                 aligned BAM file for part1,[required]
  part2                 aligned BAM file for part2,[required]

optional arguments:
  -h, --help            show this help message and exit
  -g GENOME, --genome GENOME
                        genome information, choice: mm9/mm10/hg19 et.al.,
                        [default: mm9]
  -b BIN, --bin BIN     window size for the bins for coverage calling, [default: 100000.0]
  -o OUTPUT, --output OUTPUT
                        output pdf file name, [default: Interactome_view.pdf]</pre>
</div>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">part1, part2 BAM files are the ones generated from <a class="reference internal" href="Analysis_pipeline.html#step5"><em>Step5:Stitch-seq_Aligner.py</em></a> of the pipeline; Interaction txt file is the output of <a class="reference internal" href="Analysis_pipeline.html#step6"><em>Step6:Select_strongInteraction_pp.py</em></a>.</p>
</div>
</div>
<div class="section" id="example-of-result-graph">
<h2>Example of result graph<a class="headerlink" href="#example-of-result-graph" title="Permalink to this headline">¶</a></h2>
<p>Example code:</p>
<div class="highlight-python"><pre>Rscript Plot_Circos.R GGCG_interaction_clusters.txt
  sort_Paired1_fragment_GGCG.bam sort_Paired2_fragment_GGCG.bam
  -b 100000 -o Interactome_GGCG.pdf</pre>
</div>
<p>Result figure:</p>
<img alt="_images/Interactome_GGCG.JPG" src="_images/Interactome_GGCG.JPG" />
<p>Explanation:</p>
<ul>
<li>The <font color="#763a7a"> purple </font>track right inside chromatin cytoband ideogram is the coverage of part1 (the first genomic regions connected with linker sequences)  of this sample.</li>
<li>The <font color="#0288ad"> light blue </font>track next is the coverage of part2 (the other genomic regions connected with linkers). </li>
<li>The <font color="red">inner </font>links are the strong interactions after removing rRNA. colors represent the confidence of the interaction (the ones with lower p-values are stronger) </li>
</ul></div>
</div>




  

<div id="disqus_thread"></div>
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'stitchseqtools'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'http://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="http://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>



          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Visualization of RNA-RNA interactome</a><ul>
<li><a class="reference internal" href="#prerequirement">Prerequirement</a></li>
<li><a class="reference internal" href="#run-the-program-to-generate-visualization">Run the program to generate visualization</a></li>
<li><a class="reference internal" href="#example-of-result-graph">Example of result graph</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Analysis_pipeline.html"
                        title="previous chapter">Analysis pipeline</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Other_api.html"
                        title="next chapter">Python APIs created for this project</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/Visualization.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Other_api.html" title="Python APIs created for this project"
             >next</a> |</li>
        <li class="right" >
          <a href="Analysis_pipeline.html" title="Analysis pipeline"
             >previous</a> |</li>
        <li><a href="index.html">RNA-Stitch-seq-tools 0.2.1 documentation</a> &raquo;</li> 
      </ul>
    </div>

    <div class="footer">
        &copy; Copyright 2013, Pengfei Yu.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2b2.
    </div>



<div id="editor-trap">
      <h3>Edit this document!</h3>

      <p>
        This file can be edited directly through the Web. Anyone can
        update and fix errors in this document with few clicks --
        no downloads needed.
      <p>

      <ol>

        <li>
          Go to
        <a href="https://github.com/yu68/stitch-seq/blob/master/Visualization.rst">
          Visualization of RNA-RNA interactome
        </a> on GitHub.
      </li>

        <li>
        <b>Edit</b> files using GitHub's text editor in your web browser (see the 'Edit' tab on the top right of the file)
      </li>

      <li>
        Fill in the <b>Commit message</b> text box at the bottom of the page describing <i>why</i>
        you made the changes. Press the <b>Propose file change</b> button next to it when done.
      </li>

      <li>
        Then click <b>Send a pull request</b>.
      </li>

      <li>
        Your changes are now queued for review under the project's <a href="https://github.com/yu68/stitch-seq/pulls">Pull requests</a> tab on GitHub!
      </li>
      </ol>

      <p>
        For an introduction to the documentation format please see <a href="http://docutils.sourceforge.net/docs/user/rst/quickstart.html">the reST primer</a>.
      </p>

  </div>



<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', ]);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>


  </body>
</html>